NuRadioReco/utilities/trace_utilities.py

```diff
@@ -98,18 +97,242 @@ def get_channel_voltage_from_efield(station, electric_field, channels, detector,
+    else:
+        return signal_energy_fluence
+
+def get_noise_fluence_estimators(trace, times, signal_window_mask, spacing_noise_signal=20*units.ns, relative_taper_width=0.142857143, use_median_value=False):
+    """
+    Estimate the noise fluence from the trace.
+
+    Parameters
+    ----------
+    trace : np.ndarray
+        Trace to estimate the noise fluence from.
+    times : np.ndarray
+        Time grid for the trace.
+    signal_window_mask : np.ndarray
+        Boolean mask for the signal window.
+    spacing_noise_signal : float (optional)
+        Spacing between noise windows and signal window. Makes sure no signal leaks into the noise windows.
+    relative_taper_width : float (optional)
+        Width of the taper region for the Tukey window relative to the full window length.
+    use_median_value : bool (optional)
+        If True, the median of the squared spectra of the noise windows is used as estimator. Otherwise, the mean is used.
+
+    Returns
+    -------
+    np.ndarray
+        Estimators for the noise fluence.
+    np.ndarray
+        Frequencies corresponding to the estimators.
+    """
+
+    dt = times[1] - times[0]
+    n_samples_window = sum(signal_window_mask)
+    signal_start = times[signal_window_mask][0] - spacing_noise_signal
+    signal_stop = times[signal_window_mask][-1] + spacing_noise_signal
+    list_ffts_squared = []
+
+    if signal_start < times[0]:
+        signal_start = times[0]
+        logger.warning("The signal window overlaps with the start of the trace. The efield pulse may be partialy outside the trace.")
+    elif signal_stop > times[-1]:
+        logger.warning("The signal window overlaps with the end of the trace. The efield pulse may be partialy outside the trace.")
+
+    #generate Tukey window
+    window = scipy.signal.windows.tukey(n_samples_window, relative_taper_width * 2)
+
+    #loop over the trace defining noise windows (excluding the signal window)
+    noise_start = times[0]
+    while noise_start < times[-1]:
+
+        noise_stop = noise_start + n_samples_window * dt
+        if noise_stop > times[-1]:
+            break
+
+        elif (noise_stop <= signal_start and noise_start < signal_start) or (noise_stop > signal_stop and noise_start >= signal_stop):
+
+            #clipping the noise window (rounding is needed because noise_stop = noise_start + n_samples_window * dt has numerical uncertainties)
+            mask_time = np.all([np.round(times, 5) >= np.round(noise_start, 5), np.round(times, 5) < np.round(noise_stop, 5)], axis=0)
+            time_trace_clipped = trace[mask_time]
+
+            #applying the Tukey window
+            windowed_trace = time_trace_clipped * window
 
-    if signal_window_mask is None:
-        f_signal = np.sum(electric_field_trace ** 2, axis=1)
+            #calculating the spectrum and frequencies
+            frequencies_window = np.fft.rfftfreq(len(windowed_trace), d=dt)
```

Comment:
can you use the method `fft.freqs` - does the same thing but as a plus it caches the result which should speed up things.